Add Action::ChangeAction mix-in

nweathersbee opened the pull request:
This brings in the `MuchRails::Action::ChangeAction` mix-in which
defines common behavior for upcoming `SaveAction` and `DestroyAction`
mix-ins.

# Other Changes
### Add `SaveAction` and `DestroyAction` mix-ins

This brings in `MuchRails::Action::SaveAction` and
`MuchRails::Action::DestroyAction` mix-ins. These mix-ins define
their save/destroy result, which aliases `change_result` on the
`ChangeAction` mix-in they both bring in.

nweathersbee committed:
Add Action::ChangeAction mix-in

This brings in the `MuchRails::Action::ChangeAction` mix-in which
defines common behavior for upcoming `SaveAction` and `DestroyAction`
mix-ins.

kellyredding commented on lib/much-rails.rb:
```diff
@@ -5,6 +5,15 @@
 
 require "much-rails/version"
 require "much-rails/action"
+require "much-rails/action/base_command_result"
+require "much-rails/action/base_result"
+require "much-rails/action/head_result"
+require "much-rails/action/redirect_to_result"
+require "much-rails/action/render_result"
+require "much-rails/action/save"
+require "much-rails/action/send_data_result"
+require "much-rails/action/send_file_result"
+require "much-rails/action/unprocessable_entity_result"
```
why? Shouldn't requiring `/action` do this for you?

kellyredding commented on lib/much-rails/action/save.rb:
```diff
@@ -0,0 +1,78 @@
+require "much-rails/config"
+require "much-rails/plugin"
+
+# MuchRails::Action::Save defines the common behaviors for all view
+# action classes that save records.
+module MuchRails; end
+module MuchRails::Action; end
+module MuchRails::Action::Save
```
Name is super weird. I'm thinking this should be renamed to `lib/much-rails/save_action.rb` and `MuchRails::SaveAction`.

Thoughts?

Same goes for coming `MuchRails::{Remove|Destroy}Action`s

kellyredding commented on lib/much-rails.rb:
```diff
@@ -5,6 +5,7 @@
 
 require "much-rails/version"
 require "much-rails/action"
+require "much-rails/action/change_action"
```
I wouldn't require this here. Requiring in "much-rails/action" should do it for you. So please move this to `much-rails/action`.

kellyredding commented on lib/much-rails/action/change_action.rb:
```diff
@@ -0,0 +1,78 @@
+        raise(
+          MuchRails::Action::ActionError,
+          unhandled_change_result_action_error_message,
+          # NOTE: It may be better to kick the can down the road by doing:
+          # `cause: save_result.exception` instead?
```
I think we can delete this comment, yes? It is at least stale.

kellyredding commented on lib/much-rails/action/change_action.rb:
```diff
@@ -0,0 +1,78 @@
+      @change_result ||=
+        begin
+          unless self.class.much_rails_action_change_action_config.change_result_block
+            raise(Error, "A `change_result` block must be defined.")
```
This may be future coding, but since you didn't also add SaveAction and DestroyAction here, this may need to change either here or in the future b/c the user won't actually set a `change_result` block. They will either set a `save_result` block or a `destroy_result` block:

```ruby
            raise(Error, "A change result block (e.g. `save_result` or `destroy_result` must be defined.")
```

Another approach might be to leave this alone and have the Save/Destroy Actions raise their own specific errors with specific error messages. Thoughts?

kellyredding reviewed (commented):
@nweathersbee a few comments to discuss.

nweathersbee committed:
Add `SaveAction` and `DestroyAction` mix-ins

This brings in `MuchRails::Action::SaveAction` and
`MuchRails::Action::DestroyAction` mix-ins. These mix-ins define
their save/destroy result, which aliases `change_result` on the
`ChangeAction` mix-in they both bring in.

kellyredding reviewed (approved):
@nweathersbee changes look good - thanks!
